gds_write/def2stream.py

```diff
@@ -0,0 +1,208 @@
+  )?
+  (?P<opc>                      # OPC, None if absent
+  \s+\+\ OPC
+  )?
+  \s+RECT\
+   \(\ (?P<xlo>\d+)\ (?P<ylo>\d+)\ \)\   # rect lower-left pt
+  \(\ (?P<xhi>\d+)\ (?P<yhi>\d+)\ \)\ ; # rect upper-right pt
+  ''',
+                      re.VERBOSE)
+
+def read_fills(top):
+  if args.fill_config == '':
+    print('WARNING: no fill config file specified')
+    return
+  # KLayout doesn't support FILL in DEF so we have to side load them :(
+  cfg = read_cfg()
+  in_fills = False
+  units = None
+  with open(args.input_def) as fp:
+    for line in fp:
+      if in_fills:
+        if re.match('END FILLS', line):
+          break # done with fills; don't care what follows
+        m = re.match(rect_pat, line)
+        if not m:
+          raise Exception('Unrecognized fill: ' + line)
+        opc_type = 'opc' if m.group('opc') else 'non-opc'
+        mask = m.group('mask')
+        if not mask: #uncolored just uses first entry
+          mask = 0
+        else:
+          mask = int(mask) - 1 # DEF is 1-based indexing
+        layer = cfg[m.group('layer')][opc_type]['klayout'][mask]
+        xlo = int(m.group('xlo')) / units
+        ylo = int(m.group('ylo')) / units
+        xhi = int(m.group('xhi')) / units
+        yhi = int(m.group('yhi')) / units
+        top.shapes(layer).insert(db.DBox(xlo, ylo, xhi, yhi))
+      elif re.match('FILLS \d+ ;', line):
+        in_fills = True
+      elif not units:
+        m = re.match('UNITS DISTANCE MICRONS (\d+)', line)
+        if m:
+          units = float(m.group(1))
+
+# Load technology file
+tech = db.Technology()
+tech.load(args.tech_file)
+layoutOptions = tech.load_layout_options
+if len(args.layer_map) > 0:
+  layoutOptions.lefdef_config.map_file = args.layer_map
+
+# Load def file
+main_layout = db.Layout()
+print("[INFO] Reporting cells prior to loading DEF ...")
+for i in main_layout.each_cell():
+  print("[INFO] '{0}'".format(i.name))
+
+print("[INFO] Reading DEF ...")
+main_layout.read(args.input_def, layoutOptions)
+
+#print("[INFO] Reporting cells after loading DEF ...")
+#for i in main_layout.each_cell():
+#  print("[INFO] '{0}'".format(i.name))
+
+# Clear cells
+top_cell_index = main_layout.cell(args.design_name).cell_index()
+
+# remove orphan cell BUT preserve cell with VIA_
+#  - KLayout is prepending VIA_ when reading DEF that instantiates LEF's via
+print("[INFO] Clearing cells...")
+for i in main_layout.each_cell():
+  if i.cell_index() != top_cell_index:
+    if not i.name.startswith("VIA_"):
+      i.clear()
+
+# Load in the gds to merge
+print("[INFO] Merging GDS/OAS files...")
+for fil in args.additional_gds:
+  print("\t{0}".format(fil))
+  main_layout.read(fil)
+
+# Copy the top level only to a new layout
+print("[INFO] Copying toplevel cell '{0}'".format(args.design_name))
+top_only_layout = db.Layout()
+top_only_layout.dbu = main_layout.dbu
+top = top_only_layout.create_cell(args.design_name)
+top.copy_tree(main_layout.cell(args.design_name))
```

Comment:
if we only copy the top level, did we actually need to clear the orphan cells?

Reply:
I guess it is actually unnecessary. I don't see the point of removing the orphans from `main_layout` if we copy one specific cell tree from it and then use this cell through the rest of the flow. Thank you for pointing that out.